parquet-variant/src/main/java/org/apache/parquet/variant/VariantElementConverter.java

```diff
@@ -0,0 +1,668 @@
+      builder.endObject();
+    } else if (!hasTypedValue) {
+      // There was no value or typed_value.
+      if (objectFieldName != null) {
+        // Missing field.
+        ((VariantObjectBuilder) this.holder.builder).dropLastKey();
+      } else {
+        // For arrays and top-level fields, the spec considers this invalid, but recommends using VariantNull.
+        builder.appendNull();
+      }
+    }
+  }
+
+  /**
+   * Converter for the metadata column. It sets the current metadata in the parent converter,
+   * so that it can be used by the typed_value converter on the same row.
+   */
+  static class VariantMetadataConverter extends PrimitiveConverter implements VariantConverter {
+    private VariantBuilderTopLevelHolder builder;
+    Binary[] dict;
+
+    public VariantMetadataConverter() {
+      dict = null;
+    }
+
+    @Override
+    public void init(VariantBuilderHolder builderHolder) {
+      builder = (VariantBuilderTopLevelHolder) builderHolder;
+    }
+
+    @Override
+    public void addBinary(Binary value) {
+      builder.setMetadata(value);
+    }
+
+    @Override
+    public boolean hasDictionarySupport() {
+      return true;
+    }
+
+    @Override
+    public void setDictionary(Dictionary dictionary) {
+      dict = new Binary[dictionary.getMaxId() + 1];
+      for (int i = 0; i <= dictionary.getMaxId(); i++) {
+        dict[i] = dictionary.decodeToBinary(i);
+      }
+    }
+
+    @Override
+    public void addValueFromDictionary(int dictionaryId) {
+      builder.setMetadata(dict[dictionaryId]);
+    }
+  }
+
+  // Converter for the `value` field. It does not append to VariantBuilder directly: it simply holds onto
+  // its value for the parent converter to append.
+  static class VariantValueConverter extends PrimitiveConverter implements VariantConverter {
+    private VariantElementConverter parent;
+    Binary[] dict;
+    Binary currentValue;
+
+    public VariantValueConverter(VariantElementConverter parent) {
+      this.parent = parent;
+      this.currentValue = null;
+      dict = null;
+    }
+
+    @Override
+    public void init(VariantBuilderHolder builderHolder) {}
+
+    void reset() {
+      currentValue = null;
+    }
+
+    Binary getValue() {
+      return currentValue;
+    }
+
+    @Override
+    public void addBinary(Binary value) {
+      currentValue = value;
+    }
+
+    @Override
+    public boolean hasDictionarySupport() {
+      return true;
+    }
+
+    @Override
+    public void setDictionary(Dictionary dictionary) {
+      dict = new Binary[dictionary.getMaxId() + 1];
+      for (int i = 0; i <= dictionary.getMaxId(); i++) {
+        dict[i] = dictionary.decodeToBinary(i);
+      }
+    }
+
+    @Override
+    public void addValueFromDictionary(int dictionaryId) {
+      currentValue = dict[dictionaryId];
+    }
+  }
+
+  // Base class for converting primitive typed_value fields.
+  static class VariantScalarConverter extends PrimitiveConverter implements VariantConverter {
```

Comment:
`VariantScalar` doesn't necessarily signal that this is a shredded field How about `ShreddedScalarConverter`?